### ISSUE TITLE
[Bug]: Referencing targets found only in dependent packages causes restore failure

### ISSUE THREAD

**Issue by uecasm** [NuGet/Home#11958]
### NuGet Product Used

MSBuild.exe

### Product Version

VS2022 17.2.3; msbuild 17.2.1.25201

### Worked before?

It works with classic non-PackageReference references

### Impact

It's more difficult to complete my work

### Repro Steps & Context

1. Clone https://github.com/uecasm/SassTest.git
    * This is a lightly modified Blazor Server template project, but most of the details don't really matter. 
2. Open the solution and try to build it.
3. Observe that the SassLibrary "Dependencies" node has a yellow warning icon (why no tooltip?) and in the console/error list appears;
    ```
    SassLibrary.csproj(25,73): error MSB4057: The target "LibSass_Build" does not exist in the project.
    ```
4. If you look at the SassLibrary's `projects.assets.json` or `.nuget.g.props` (etc) files now, they are mostly empty and do not report any dependencies.  (In addition, while in this state the "manage packages for project" view reports no installed packages.)
5. Open the library project file, comment out the entire `BuildScopedSass` target, and then save the project file.
6. The warning icons go away; build now and it succeeds (but builds something at the wrong time).
7. Uncomment the target and save; the warning icons do not return and building still succeeds (and now it does the build step properly).
8. Closing/reopening/rebuilding the solution etc will not reintroduce the problem.  To "re-arm" this bug, **both** the `.vs` directory and the `SassLibrary\obj` directory (or presumably some subset of files within) need to be deleted while the solution is not open; the bug will then reoccur after opening the solution again.  But this will of course happen whenever there's a clean checkout, e.g. for CI or on a new developer's machine.

The target in question is provided by the referenced package, which is why this works after the package restore is actually completed.  I'm not entirely sure why this continues to work even if only the `obj` folder and not the `.vs` folder is deleted, but that does appear to be the case.

Pre-`PackageReference`, you could solve this by referencing the additional target only after it was imported.  However there is no clear way to do this with the implied imports.  (Similarly, it is unclear when or if it might be safe to define a dummy target in the project, expecting that it will be overridden by the import "later".)  Ideally, msbuild should ignore these sorts of errors when performing package restore (and complete the restore successfully), since it doesn't need to execute those targets anyway.

----
While testing permutations for this bug report, I've managed to successfully work around the problem by moving the target reference from `DependsOnTargets` to `CallTargets` within the target body.  However this feels like a kludge (and is not entirely equivalent), so I still think the underlying problem is a bug.

**Comment by jeffkl** [comment]
@uecasm Thanks for reporting this issue.  I believe its a known problem with how Visual Studio collects PackageReference items before restore.  This is done by the [Project System](https://github.com/dotnet/project-system) in a [design-time build](https://github.com/dotnet/project-system/blob/main/docs/design-time-builds.md#determining-whether-a-target-is-running-in-a-design-time-build).  This build is failing because your custom target is trying to depend on a target that does not exist.  Since the design-time build fails, NuGet does not receive a complete list the PackageReference items and does not restore anything.  The workaround is to ensure that custom targets like this do not run during design-time builds:

```diff
- <Target Name="BuildScopedSass" BeforeTargets="ResolveScopedCssInputs" DependsOnTargets="LibSass_Build">
+ <Target Name="BuildScopedSass" BeforeTargets="ResolveScopedCssInputs" DependsOnTargets="LibSass_Build" Condition="'$(DesignTimeBuild)' != 'true'">
```

**Comment by jeffkl** [comment]
Closing this in favor of https://github.com/dotnet/project-system/issues/8318
